src/main/java/net/minecraftforge/gui/capability/IGuiProvider.java

```diff
@@ -0,0 +1,61 @@
+package net.minecraftforge.gui.capability;
+
+import io.netty.buffer.ByteBuf;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.inventory.Container;
+import net.minecraft.world.World;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents a gui provider- something that activates a gui for a player to use.
+ */
+public interface IGuiProvider<T>
+{
+    /**
+     * Deserializes a buffer from the server into the provider instance.
+     *
+     * @param buffer
+     */
+    void deserialize(ByteBuf buffer);
+
+    /**
+     * Serializes data on the server to send along with the open gui packet.
+     * Use this to add any information you need to open the gui properly on the client.
+     *
+     * @param buffer
+     */
+    void serialize(ByteBuf buffer);
+
+    /**
+     * Gets the gui provider object from the world, player, etc.
+     * Used in the get*Element methods to pass a direct instance to them.
+     *
+     * @param player The player instance.
+     * @param world The world.
+     * @return
+     */
+    T getOwner(EntityPlayer player, World world);
+
+    /**
+     * Gets a client-side gui element (may be null) to display to a player.
+     *
+     * @param player The player viewing the gui.
+     * @param world The world the gui is being displayed in.
+     * @param owner The owner object of the gui. Usually a tile, item, or entity.
+     * @return A Gui/GuiScreen to display on the client.
+     */
+    @Nullable
+    Object getClientGuiElement(EntityPlayer player, World world, @Nullable T owner);
+
+    /**
+     * Gets a server-side container to handle slots and interaction on the server.
+     *
+     * @param player The player viewing the gui.
+     * @param world The world the gui is being accessed on.
+     * @param owner The owner object. Usually a tile, item, or entity.
```

Comment:
Please explain the `@Nullable` by describing the conditions where owner can be null